```{r inv-1301, echo=FALSE}
    startQuest("Biological QA - Variance Stabilising Transformation")
```

```{r head-1301, echo=FALSE}
    h1("Sexual dimorphism study Biological QA")
```

```{r message=FALSE,warning=FALSE,results='hide'}
    library(RnaSeqTutorial)
```

```{r head-1302, echo=FALSE}
    h2("(Re)Introduction")
```
As a running example, we use a dataset derived from a study performed
in Populus tremula (Robinson, Delhomme et al., BMC Plant Biology, 2014}. 
The authors test the evolutionary theory suggesting
that males and females may evolve sexually dimorphic phenotypic and biochemical 
traits concordant with each sex having different optimal strategies of resource 
investment to maximise reproductive success and fitness. Such sexual dimorphism
should result in sex biased gene expression patterns in non-floral organs for
autosomal genes associated with the control and development of such phenotypic 
traits. Hence, the authors, among other approaches have looked for gene expression 
differences between sex. This was achieved by an RNA-Seq differential expression 
analysis between samples grouped by sex. The samples have been sequenced on an
Illumina HiSeq 2000, using TruSeq adapters, through a 101 cycles paired-end protocol,
which yielded between 11 million and 34 million reads per sample. 

```{r head-1303, echo=FALSE}
    h2("Processing the data")
```

```{r head-1304, echo =FALSE}
    h3("Reading in the data")
```

First we read the HTSeq files in a matrix. The DESeq2 package now
actually has a function to ease that process: DESeqDataSetFromHTSeqCount. 
Here we just process the samples in parallel using mclapply instead.

```{r }
res <- mclapply(dir(file.path(extdata(),"htseq"),
                    pattern="^[2,3].*_STAR\\.txt",
                    full.names=TRUE),function(fil){
  read.delim(fil,header=FALSE,stringsAsFactors=FALSE)
},mc.cores=16)
names(res) <- gsub("_.*_STAR\\.txt","",dir(file.path(extdata(),"htseq"),
                                           pattern="^[2,3].*_STAR\\.txt"))
```

Then we extract the additional info that HTSeq writes at the end of every
file detailing how many reads are aligned uniquely or not, have no or ambigous
overlaps or are of too low a quality. That last category should not occur at this
stage of the analysis, as we have sorted and trimmed the data before aligning.

```{r }
addInfo <- c("__no_feature","__ambiguous",
             "__too_low_aQual","__not_aligned",
             "__alignment_not_unique")
```

and we then extract the reads

```{r }
sel <- match(addInfo,res[[1]][,1])
count.table <- do.call(cbind,lapply(res,"[",2))[-sel,]
colnames(count.table) <- names(res)
rownames(count.table) <- res[[1]][,1][-sel]
```

```{r head-1305, echo =FALSE}
     h3("The HTSeq stat lines")
```

Here we aggreagte the information about how many
reads aligned, how many were ambiguous, etc

```{r }
count.stats <- do.call(cbind,lapply(res,"[",2))[sel,]
colnames(count.stats) <- names(res)
rownames(count.stats) <- sub("__","",addInfo)
count.stats <- rbind(count.stats,aligned=colSums(count.table))
count.stats <- count.stats[rowSums(count.stats) > 0,]
```

Then we convert them as percentages and check them all

```{r }
apply(count.stats,2,function(co){round(co*100/sum(co))})
```

As can be seen an average 82% of the reads aligned uniquely
unambiguously to features. About 13% were mapping multiple 
locations, 2% were on ambigous features (as the data is non
strand specific, we have used the "Union" counting scheme) and
finally 3% align to no features. This is expected
given the fact that the alignments were done against the 
the P. trichocarpa genome.

```{r head-1306, echo =FALSE}
     h3("Visualizing the HTSeq stats")
```

```{r }
pal=brewer.pal(6,"Dark2")[1:nrow(count.stats)]
mar <- par("mar")
par(mar=c(7.1,5.1,4.1,2.1))
barplot(as.matrix(count.stats),col=pal,beside=TRUE,las=2,main="read proportion",
        ylim=range(count.stats)+c(0,1e+7))
legend("top",fill=pal,legend=rownames(count.stats),bty="n",cex=0.8,horiz=TRUE)
par(mar=mar)
```

Here, one can clearly see a library size effect; the samples 229.1 and 349.2
have been sequenced deeper as these are the parents of a large scale F1 population
established for another study. Appart from these, the amount of reads per library
seems fairly equally distributed.

```{r head-1307, echo =FALSE}
     h3("Deriving more information")
```

We can estimate how much of the genes are not expressed

```{r }
sel <- rowSums(count.table) == 0
sprintf("%s percent",round(sum(sel) * 100/ nrow(count.table),digits=1))
sprintf("of %s genes are not expressed",nrow(count.table))
```

So 13.9% of the genes are not expressed out of a total of 41335 genes

```{r head-1309, echo =FALSE}
     h2("Biological QA")
```

To assess the validity of the replicates, we can look at
different metrics. 

```{r head-1310, echo =FALSE}
     h3("Raw count distribution")
```

We start by looking at the
per-gene mean expression, i.e. the mean raw count of every 
gene across samples is calculated and displayed on a log10 scale

```{r }
plot(density(log10(rowMeans(count.table))),col=pal[1],
     main="mean raw counts distribution",
     xlab="mean raw counts (log10)")
```

Then the same is done for the individual samples colored by sample type

```{r }
pal=brewer.pal(8,"Dark2")
plot.multidensity(log10(count.table),col=rep(pal,each=3),
                  legend.x="topright",legend.cex=0.5,
                  main="sample raw counts distribution",
                  xlab="per gene raw counts (log10)")
```

As one can see, most samples show the similar trends, a few samples are
shifted to the right - those that were sequenced deeper, but this does not
affect the global shape of the curve.

```{r head-1311, echo =FALSE}
     h2("Variance stabilisation for better visualisation")
```

For visualization, the data is submitted to a variance 
stabilization transformation using DESeq2. The 
 dispersion is going to be estimated independently of the sample type but
we nonetheless define the metadata fo importance, i.e. every
sample sex and year of sampling

```{r }
samples <- read.csv(file.path(extdata(),"sex-samples.csv"))
sex <- samples$sex[match(names(res),samples$sample)]
date <- factor(samples$date[match(names(res),samples$sample)])
```

And then we create the DESeq2 object, using the sample name
as condition (which is hence irrelevant to any comparison)

```{r }
dds <- DESeqDataSetFromMatrix(
  countData = count.table,
  colData = data.frame(
                       sex=sex,
                       date=date),
  design = ~ date + sex)
```

Once this is done, we first check the size factors and
as there is no big variation, we decide to go for a variance
stabilisation transformation over a regularised log transformation.
Check the DESeq2 vignette for more details about this.

```{r }
dds <- estimateSizeFactors(dds)
sizes <- sizeFactors(dds)
names(sizes) <- names(res)
sizes
boxplot(sizes,main="relative library sizes",ylab="scaling factor")
```

Let us do the vst. Here we do not want to give any prior to the normalisation,
so we perform a __BLIND__ transformation. If we were planning to work with the 
_vst_ data for non-QC analysis, then we __WOULD__ set `blind=FALSE` to ensure
a more adequate dispersion estimation.

```{r }
vsd <- varianceStabilizingTransformation(dds, blind=TRUE)
vst <- assay(vsd)
colnames(vst) <- colnames(count.table)
```

The vst introduces an offset, i.e. non expressed gene will have a value set
as the minimal value of the vst. To avoid lengthy unnecessary explanations, we
simply shift all values so that non expressed genes have a vst value of 0

```{r }
vst <- vst - min(vst)
```

It is then essential to validate the vst effect and ensure its validity. To
that extend, we visualize the corrected mean - sd relationship. As it is 
fairly linear, meaning we can assume homoscedasticity.
the slight initial trend / bump is due to genes having few 
counts in a few subset of the samples and hence having a 
higher variability. This is expected.

```{r }
meanSdPlot(assay(vsd)[rowSums(count.table)>0,])
```

Here is what would be observed by a log2 transformation of the raw count data

```{r }
meanSdPlot(log2(as.matrix(count.table[rowSums(count.table)>0,])))
```

or for the count adjusted for the library size factor

```{r }
meanSdPlot(log2(counts(dds,normalized=TRUE)[rowSums(count.table)>0,]))
```

```{r inv-1302, echo=FALSE}
    quest(2)
    endQuest()
    startQuest("Biological QA - PCA and heatmap")
```

```{r head-1312, echo =FALSE}
     h3("PCA analysis")
```

We perform a Principal Component Analysis (PCA) of the data
to do a quick quality assessment, i.e. replicate should 
cluster and the first dimensions should be explainable by 
biological means.

```{r }
pc <- prcomp(t(vst))
percent <- round(summary(pc)$importance[2,]*100)
smpls <- conditions
```

We color the samples by date and sex

```{r }
sex.cols<-c("pink","lightblue")
sex.names<-c(F="Female",M="Male")
```

And then check wheter the observed separation is due to the sample sex

```{r }
scatterplot3d(pc$x[,1],
              pc$x[,2],
              pc$x[,3],
              xlab=paste("Comp. 1 (",percent[1],"%)",sep=""),
              ylab=paste("Comp. 2 (",percent[2],"%)",sep=""),
              zlab=paste("Comp. 3 (",percent[3],"%)",sep=""),
              color=sex.cols[as.integer(factor(sex))],
              pch=19)
legend("bottomright",pch=c(NA,15,15),col=c(NA,sex.cols[1:2]),
       legend=c("Color:",sex.names[levels(factor(sex))]))
par(mar=mar)
```

This does not seem to be the case at all. There are 2 clusters of points, but
both equally contain pink and blue dots. So, we next color by sampling date

```{r }
dat <- date
scatterplot3d(pc$x[,1],
              pc$x[,2],
              pc$x[,3],
              xlab=paste("Comp. 1 (",percent[1],"%)",sep=""),
              ylab=paste("Comp. 2 (",percent[2],"%)",sep=""),
              zlab=paste("Comp. 3 (",percent[3],"%)",sep=""),
              color=pal[as.integer(factor(dat))],
              pch=19)
legend("topleft",pch=rep(c(19,23),each=10),col=rep(pal,2),legend=levels(factor(dat)),bty="n")
par(mar=mar)
```

And here we are... The sampling data is a STRONG CONFOUNDING FACTOR in our
analysis. So let's do a final plot with the color = sex and shape = date

```{r }
scatterplot3d(pc$x[,1],
              pc$x[,2],
              pc$x[,3],
              xlab=paste("Comp. 1 (",percent[1],"%)",sep=""),
              ylab=paste("Comp. 2 (",percent[2],"%)",sep=""),
              zlab=paste("Comp. 3 (",percent[3],"%)",sep=""),
              color=sex.cols[as.integer(factor(sex))],
              pch=c(19,17)[as.integer(factor(dat))])
legend("bottomright",pch=c(NA,15,15),col=c(NA,sex.cols[1:2]),
       legend=c("Color:",sex.names[levels(factor(sex))]))
legend("topleft",pch=c(NA,21,24),col=c(NA,1,1),
       legend=c("Symbol:",levels(factor(dat))),cex=0.85)
par(mar=mar)
```

Sometimes the 3D PCA are not so easy to read and we may be just
as well off looking at 2 dimensions at a time. First the 2 first dims

```{r }
plot(pc$x[,1],
     pc$x[,2],
     xlab=paste("Comp. 1 (",percent[1],"%)",sep=""),
     ylab=paste("Comp. 2 (",percent[2],"%)",sep=""),
     col=sex.cols[as.integer(factor(sex))],
     pch=c(19,17)[as.integer(factor(dat))],
     main="Principal Component Analysis",sub="variance stabilized counts")
legend("bottomleft",pch=c(NA,15,15),col=c(NA,sex.cols[1:2]),
       legend=c("Color:",sex.names[levels(factor(sex))]))
legend("topright",pch=c(NA,21,24),col=c(NA,1,1),
       legend=c("Symbol:",levels(factor(dat))),cex=0.85)
text(pc$x[,1],  
     pc$x[,2],
     labels=colnames(count.table),cex=.5,adj=-1)
```

Clearly the sampling date separate most samples. It is interesting to see that
the 2 samples sequenced deeper are in between the 2 clusters. We could hypothesize
that the environmental change between sampling years are affecting less important 
genes, hence genes which expression levels are lower; i.e. somewhat looking at 
transcriptional "noise". Since the 2 parent samples have been sequenced much deeper, 
the fact that they group with neither cluster might be due to that fact that they too have 
an increased proportion of transcriptional noise.
Looking at the 2nd and 3rd dims does not reveal any obvious effects

```{r }
plot(pc$x[,2],
     pc$x[,3],
     xlab=paste("Comp. 2 (",percent[2],"%)",sep=""),
     ylab=paste("Comp. 3 (",percent[3],"%)",sep=""),
     col=sex.cols[as.integer(factor(sex))],
     pch=c(19,17)[as.integer(factor(dat))],     
     main="Principal Component Analysis",sub="variance stabilized counts")
legend("bottomleft",pch=c(NA,15,15),col=c(NA,sex.cols[1:2]),
       legend=c("Color:",sex.names[levels(factor(sex))]))
legend("topright",pch=c(NA,21,24),col=c(NA,1,1),
       legend=c("Symbol:",levels(factor(dat))),cex=0.85)
```

```{r inv-1303, echo=FALSE}
    quest(3)
    endQuest()
    startQuest("Kallisto vs. HTSeq")
```

```{r head-1313, echo =FALSE}
     h1("Kallisto - HTSeq comparison")
```

__Caveats:__ 
1. This section has not been properly tested, just felt like a nice
addon :-)
2. We have created the kallisto index using the synthetic transcripts, hence
we only have a one to one mapping, which defeats the purpose of running kallisto

Anyway, before moving on to the topic of differential expression, let us take a 
look at the differences between the HTSeq and Kallisto counts.

First, we read in the kallisto abundance files using the `tximport` package
```{r, eval=FALSE}
    library(tximport)
    files <- list.files("~/kallisto", 
                    pattern = ".*-abundance.tsv",
                    full.names = TRUE)
    tx <- suppressMessages(tximport(files = files,
                                     type = "kallisto", 
                                     txOut = TRUE))
```

Next, we extract the gene expression counts (since we have a 1to1 relationship)

```{r, eval=FALSE}
k.countTable <- round(tx$counts)
```

Finally, let's plot all samples kallisto vs. htseq counts.

```{r, eval=FALSE}
library(LSD)
par(mfrow=c(3,3))
dev.null <- sapply(1:ncol(count.table),function(i){
    heatscatter(log2(count.table[,i]+1),
                log2(k.countTable[,i]+1),
                main=colnames(count.table)[i],
                xlab="HTSeq count (log2 + pseudo-count)",
                ylab="Kallisto count (log2 + pseudo-count)",
                )
})
```

```{r inv-1304, echo=FALSE}
    quest(4)
    endQuest()
```
